src/meta/src/barrier/mod.rs

```diff
@@ -827,42 +903,92 @@ impl GlobalBarrierManager {
                     tracker.stash_command_to_finish(command);
                 }
 
-                if let Some(table_id) = node.command_ctx.table_to_cancel() {
+                if let Some(table_id) = command_ctx.table_to_cancel() {
                     // the cancelled command is possibly stashed in `finished_commands` and waiting
                     // for checkpoint, we should also clear it.
                     tracker.cancel_command(table_id);
                 }
 
-                let remaining = tracker.finish_jobs(kind.is_checkpoint()).await?;
-                // If there are remaining commands (that requires checkpoint to finish), we force
-                // the next barrier to be a checkpoint.
-                if remaining {
-                    assert_matches!(kind, BarrierKind::Barrier);
-                    self.scheduled_barriers.force_checkpoint_in_next_barrier();
-                }
+                let has_remaining_job = tracker
+                    .finish_jobs(command_ctx.kind.is_checkpoint())
+                    .await?;
 
-                let duration_sec = node.timer.take().unwrap().stop_and_record();
-                node.wait_commit_timer.take().unwrap().observe_duration();
+                Ok(has_remaining_job)
+            }
+        }
+    }
 
+    fn report_complete_event(&self, duration_sec: f64, command_ctx: &CommandContext) {
+        {
+            {
                 {
                     // Record barrier latency in event log.
                     use risingwave_pb::meta::event_log;
                     let event = event_log::EventBarrierComplete {
-                        prev_epoch: node.command_ctx.prev_epoch.value().0,
-                        cur_epoch: node.command_ctx.curr_epoch.value().0,
+                        prev_epoch: command_ctx.prev_epoch.value().0,
+                        cur_epoch: command_ctx.curr_epoch.value().0,
                         duration_sec,
-                        command: node.command_ctx.command.to_string(),
-                        barrier_kind: node.command_ctx.kind.as_str_name().to_string(),
+                        command: command_ctx.command.to_string(),
+                        barrier_kind: command_ctx.kind.as_str_name().to_string(),
                     };
                     self.env
                         .event_log_manager_ref()
                         .add_event_logs(vec![event_log::Event::BarrierComplete(event)]);
                 }
+            }
+        }
+    }
+}
 
-                Ok(())
+impl CheckpointControl {
+    pub(super) fn next_completed_barrier(
+        &mut self,
+    ) -> impl Future<Output = MetaResult<(Arc<CommandContext>, bool)>> + '_ {
```

Comment:
Any reason for not implementing as an async function? 👀

Reply:
The future is used in a select!, and we should make the future cancellation safe. Here we need to poll the join handle. If we implement it as a async method, we need to take the join handle out and await on it, which makes the future cancellation unsafe.

Reply:
`&mut Future` is also a `Future`. Does it work in this case?

https://doc.rust-lang.org/stable/std/future/trait.Future.html#impl-Future-for-%26mut+F

Reply:
Oops. Surprised that it works. I used to think that the `.await` sugar must take the ownership of the future, but it seems that it doesn't. 

So now we can do things like await on a oneshot rx for multiple times and pass the compile, which is somehow counterintuitive. 
```
#[tokio::main]
async fn main() {
    let (tx, mut rx) = tokio::sync::oneshot::channel();
    tx.send(()).unwrap();

    let rx_mut = &mut rx;
    rx_mut.await.unwrap();

    let rx_mut = &mut rx;
    rx_mut.await.unwrap();
}

thread 'main' panicked at /playground/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.36.0/src/sync/oneshot.rs:1109:13:
called after complete
```

Reply:
Yes. This is what `FusedFuture` for.

I believe the manual `poll` style has the same issue. It works now because there's no another point of `return Pending` between `join_handle` gets ready and replace `self.completing_command` to drop the future. This is exactly corresponding to no `yield` point after awaiting on `&mut join_handle`.